R/install_torsten.R

```diff
@@ -0,0 +1,197 @@
+#'   the `make/local` file. For example, `list("CXX" = "clang++")` will force
+#'   the use of clang for compilation.
+#' @param check_toolchain (logical) Should `install_torsten()` attempt to check
+#'   that the required toolchain is installed and properly configured. The
+#'   default is `TRUE`.
+#'
+#' @examples
+#' \dontrun{
+#' # install_torsten(cores = 4)
+#'
+#' cpp_options <- list(
+#'   "CXX" = "clang++",
+#'   "CXXFLAGS+= -march=native",
+#'   PRECOMPILED_HEADERS = TRUE
+#' )
+#' # install_torsten(cores = 4, cpp_options = cpp_options)
+#' }
+#'
+
+#'## install_torsten version without unexported dependencies
+
+latest_torsten_release <- function() {
+  require(jsonlite)
+  dest_file <- tempfile(pattern = "releases-", fileext = ".json")
+  download_url <- "https://api.github.com/repos/metrumresearchgroup/Torsten/releases/latest"
+  release_list_downloaded <- utils::download.file(url = download_url,
+                                                  destfile = dest_file,
+                                                  quiet = TRUE)
+  if (release_list_downloaded != 0) {
+    stop("GitHub download of release list failed.", call. = FALSE)
+  }
+  release <- jsonlite::read_json(dest_file)
+  release$tag_name
+}
+
+get_torsten_release_list <- function() {
+  require(jsonlite)
+  dest_file <- tempfile(pattern = "releases-", fileext = ".json")
+  download_url <- "https://api.github.com/repos/metrumresearchgroup/Torsten/releases"
+  release_list_downloaded <- utils::download.file(url = download_url,
+                                                  destfile = dest_file,
+                                                  quiet = TRUE)
+  if (release_list_downloaded != 0) {
+    stop("GitHub download of release list failed.", call. = FALSE)
+  }
+  release <- jsonlite::read_json(dest_file)
+  sapply(release,
+         function(x){
+           x$tag_name
+         })
+}
+
+install_torsten <- function(dir = NULL,
+                            cores = getOption("mc.cores", 2),
+                            quiet = FALSE,
+                            timeout = 1200,
+                            version = NULL,
+                            release_url = NULL,
+                            cpp_options = list(),
+                            check_toolchain = TRUE
+) {
+
+  ## install_torsten is based on cmdstanr::install_cmdstan. Differences include
+  ## * install_torsten always overwrites pre-existing Torsten installation at the
+  ##.  specified cmdstan_path.
+  ## * install_torsten does not try to automatically append to cpp_options for
+  ##.  M1 Macs. Is this needed?
+  ## * install_torsten does not provide support for WSL. Should it?
+
+  require(cmdstanr)
+  if (check_toolchain) {
+    cmdstanr::check_cmdstan_toolchain(fix = FALSE, quiet = quiet)
+  }
+  if (is.null(dir)) {
+    dir <- file.path(dirname(cmdstanr::cmdstan_default_install_path()),
+                     ".torsten")
```

Comment:
I'd like to avoid littering `$HOME` with another app directory. I know cmdstanr does it, but I'm not convinced we should follow suit.

My preference would be to put these under `XDG_DATA_HOME` (e.g., `file.path(tools::R_user_dir("bbr.bayes"), "torsten")` or maybe just `file.path(tools::R_user_dir("torsten"))` ). `tools::R_user_dir()` isn't available until R 4.0, but that's okay for the purposes of bbr.bayes.

Reply:
I set the default to `file.path(tools::R_user_dir("torsten"))` and revised the description accordingly.